### ISSUE TITLE
Single Log-Out Does Not Appear to Support Mupltile SPs

### ISSUE THREAD

**Issue by mvastola** [saml-idp/saml_idp#82]
Single Log-Out was implemented (at least in theory) by #41 by @pkarman, but the purpose of the feature, (at least as I understand it) is that a logout command sent to the IdP results in logouts at any SPs that have been logged into during the login session. (See http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0-cd-02.html#5.3.2.SP-Initiated%20Single%20Logout%20with%20Multiple%20SPs|outline)

From looking at the code here though, I'm not quite sure how that's implemented. For instance, the IdP doesn't seem to maintain (or seek to use) a per-session list of the SPs that have been signed into, nor is there any sort of loop in the `SamlIdp::IdpController`. Is this support missing, or am *I* missing something?

If the former, could this be considered a feature request?

I understand that the `README.md` states:
> This was originally setup by @lawrencepit to test SAML Clients. I took it closer to a real SAML IDP implementation.

.. which implies this implementation isn't feature complete. I'm frankly quite grateful that this gem exists to begin with, but is there any way the `README.md` could also perhaps list the *ways* in which this isn't a full implementation? (It would just make it a lot easier for users to work around the incomplete features and perhaps even know exactly what to offer PRs for.)

Thanks a ton,
Mike

**Comment by pkarman** [comment]
@mvastola the functionality I added to this library was the basic request/response handling to mirror what was done in the ruby-saml gem.

login.gov uses the functionality I added in the this project:
https://github.com/18F/identity-idp/blob/master/app/controllers/concerns/saml_idp_logout_concern.rb

the loop you're referring to is implemented here:
https://github.com/18F/identity-idp/blob/master/app/services/single_logout_handler.rb#L38

where it keeps generating request/responses until all active "identity" records are exhausted.

HTH

**Comment by mvastola** [comment]
Ah. I'll have to dig through all the code in https://github.com/18F/identity-idp/, but I suspect this will immensely helpful in terms of providing a sample implementation for me. Thank you!

I am noticing though that your project uses a [fork](https://github.com/18F/saml_idp) of this gem. Are there any relevant differences between your fork and this one that impact this functionality? 

I'm going to leave this ticket open as it would be awesome to see if this feature could be implemented in this gem, or at the very least an example of how to implement it could make its way into the `README.md`.

**Comment by jphenow** [comment]
This is not actively maintained. I developed this gem while at Sport Ngin (see organization that owns this) and try to merge or respond when available, though that's not terribly frequently as you can probably see. To that end, pull requests welcome - I will try to tend to them as I'm able.

**Comment by mvastola** [comment]
Is the other one more actively maintained / a better alternative?

**Comment by pkarman** [comment]
The https://github.com/18f/saml_idp fork is actively maintained because it is a core dependency of login.gov, but it is not actively synced from this upstream version. So "better" depends on your POV.

There seems to be some community interest in this gem even if the owning org doesn't maintain it. I'm not at 18F anymore but perhaps there would be interest there in pulling in upstream PRs and taking over "ownership" of the gem.

@jphenow I think the main thing there would be giving others at 18F the ability to push releases to rubygems.org?

@monfresh @zachmargolis - thoughts on this?

**Comment by mvastola** [comment]
To me, the ideal case is a gem that's active, accepting user contributions, and on rubygems (either under this name or another one). 

I also note that the 18F fork is both behind *and* ahead of this gem, so its unfortunate in that there's bound to be some duplication of work as long as the 18F fork has features not in upstream and 18F isn't making PRs and the maintainer of this gem isn't able to readily merge them. (Obviously this all becomes moot if 18F takes over in some fashion.)

**Comment by monfresh** [comment]
I can't guarantee anything, but for now, as long as login.gov still exists, it will most likely continue to depend on this gem. As long as I'm still working on login.gov, I'd be happy to accept user contributions. If @jphenow can give us at 18F the ability to push releases to rubygems, that would be great. @mvastola Feel free to take a stab at submitting the missing commits from this gem into the 18F fork.

**Comment by mvastola** [comment]
@monfresh, ok. Will make the PRs.

As for sharing the ability to push releases, as an outside observer not at all involved in this decision, I can forsee some potential for complications here. (i.e. One org currently owns the gem name and might not want to let go. Even if they do, maintenance of the new gem becomes contingent on the whims of the government and/or one person's employment.) 

Might it be simpler to just create a new (neutral) GitHub organization and repo for this gem and either transfer ownership of the rubygem ownership to that or create a new gem name? 

That way the gem can survive @monfresh's tenure at his current job and even accumulate additional co-maintainers that aren't strictly linked to login.gov and/or 18F. In order to work well, this would of course require 18F to be willing to either use a gem not under their complete control, or begin to actively merge changes upstream.

Thoughts?

**Comment by karpet** [comment]
@mvastola I agree that a new org owning and maintaining the gem would be the ideal.

To be clear, 18F has actively merged changes upstream, until the current owners stopped responding. So there is both will and effort from the 18F side at maintaining the gem, and I imagine folks there would be happy not to have to maintain a fork if they don't have to (as they use 100s of gems that they do not maintain).

@jphenow are you on board with this idea of a new org? Can you broker the rubygems permissions required with the existing https://rubygems.org/gems/saml_idp and expand to include new releasers?

**Comment by jphenow** [comment]
I do not believe I have the rights to do this on a whim. Someone from Sports Engine should be involved. @natejgreene @anfleene - wanna speak up? Or @ someone who can?

Having been outside the organization for 2 years now, I try to maintain this, within reason, while not introducing breaking things. I've gone for a somewhat minimal approach, and I'd like to keep it that way, so whatever Sports Engine and you guys decide, I'm on board for. Having done the majority of this work while on Sports Engine payroll, I don't think it's my choice where the gem ownership or future development go.

**Comment by mvastola** [comment]
@karpet ah, ok. That makes things much simpler then if 18F has been willing to do this.

Then it would seem the only remaining issue is transferring permissions. The only other possible complication I can think of is I'm not sure if that's solely @jphenow's call or if he will feel obliged to consult with Sport Ngin, which could delay things.

I'm happy to make the new org. Which repo would be moved into it? the 18F one or this one?

**Comment by mvastola** [comment]
@jphenow yeah. Thought that might be the case. Technically, as this is under the MIT License, future development can continue anywhere we want it to. Access to this specific gem name are a separate issue.

I just created the org named https://github.com/saml-idp (underscores aren't allowed in org names). 

@monfresh, I'll make you an owner. Maybe it's best to transfer the 18F repo into the new org since there's some red tape to deal with regarding the Sports Engine one?

@karpet, @jphenow, are you interested in helping to maintain the new repo? 

@monfresh, feel free to give access to whomever needs it at 18F if you are going to move in the repo.

**Comment by karpet** [comment]
@mvastola you're right about the IP issues.

What I was mostly concerned about was the rubygems.org publishing permissions. That's where Sports Engine needs to weigh in as the owner of the gem on that site.

I suspect that 18F will want to maintain their own fork and create upstream PRs from there. Picking which repo is used to seed the new org should be up to the maintainers.

I don't think I should be a maintainer - I'm out of this problem space and am not much good for anything other than chiming in here on logistical details.

**Comment by mvastola** [comment]
@karpet, Got it. According to [RubyGems](https://rubygems.org/gems/saml_idp) the official co-owners of the gem are [jphenow](https://rubygems.org/profiles/jphenow) and [sportngin](https://rubygems.org/profiles/sportngin). 

@jphenow do you know who is behind the RubyGems username 'sportngin' (it's linked to the email infrastructure@sportngin.com, if that helps at all)? Worst case scenario is we can try emailing them and point them to this issue.

As for which repo to use, I'm hovering back and forth.. This one would be ideal, if we're going to preserve the gem name and can move the repo rather than copy it, as the `18F` one has no issues and just one PR. It also makes more sense in terms of making incremental changes from the most recent version. Thoughts, @monfresh?

**Comment by mvastola** [comment]
Any news on this?

As promised, a week ago, I made a PR into the `18F` downstream to merge almost all the more recent commits from this upstream: 18F/saml_idp#13.

**Comment by mvastola** [comment]
@monfresh I know you're going to review the PR when you get a chance, but would it make sense to discuss the logistics of the new repo off this issue via email? (Mine is public in my profile if you want to shoot me one. Among other logistics, I think we should shoot infrastructure@sportngin.com an email, but I think that should be a joint effort.

**Comment by anfleene** [comment]
@mvastola @karpet How can I help? This gem is in need of maintainers I'm happy to help make that happen.

**Comment by mvastola** [comment]
@anfleene, since it appears no one at Sports Engine is interested in maintaining this, would you be willing to relinquish this gem to a new group of maintainers? (Currently me and @monfresh, I believe.) In this way the gem would no longer be dependent on any one company for existence and also have active maintainers.

If you are okay with this, we could use your help with up to two things:
1. At a minimum what would be needed is transfer of the ownership of the [`saml_idp` gem on RubyGems](https://rubygems.org/gems/saml_idp) to me and/or @monfresh (so that we can push releases to it). It looks like the user with access currently has the username (sportngin)[https://rubygems.org/profiles/sportngin] and the email infrastructure@sportngin.com. 
*My* username on RubyGems is [mvastola](https://rubygems.org/profiles/mvastola), if you could add that one, then either you or I can add @monfresh's once we confirm his username). 
@monfresh, [monfresh](https://rubygems.org/profiles/monfresh) appears to be you. Is this both yours *and* the one you'd like to use?
2. Additionally, it would be great if we could transfer this entire repository over to the new organization I made for the `saml_idp` gem. You would do this by going to the very bottom of https://github.com/sportngin/saml_idp/settings. Inside the "Danger Zone", there is a button labeled "Transfer". You would want to click that and transfer the repository to the organization named [`saml-idp`](https://github.com/saml-idp) (note the hyphen rather than an underscore here). We will then have to approve it on our end before the transfer completes.
This step will allow us to preserve all outstanding pull requests and issues and other GitHub metadata that can't easily be copied over by simply using `git` operations.

If you have any questions or run into any issues, don't hesitate to follow up here or shoot me an email or whatever.

**Comment by anfleene** [comment]
1 is done. 2 threw an error `You don’t have the permission to create repositories on saml-idp` Can you give me permissions to make repos on saml-idp?

**Comment by mvastola** [comment]
Thanks!
Oops. Didn't know I needed to do that. Just invited you though, so you need to accept first (at https://github.com/saml-idp) and then you should be set to do step 2. Let me know if that doesn't work.

**Comment by anfleene** [comment]
Done! Happy Maintaining 👍

**Comment by mvastola** [comment]
@anfleene, thanks!!

**Comment by mvastola** [comment]
@monfresh, shoot me a line? I think it makes sense to close this and stop using this issue to discuss management.
